book/src/dev/rfcs/drafts/xxxx-async-rust.md

```diff
@@ -0,0 +1,443 @@
+
+Select's bias can be useful to ensure that cancel oneshots and timers are always
+executed first. Consider the `select_biased!` macro and `FuturesOrdered` stream
+for guaranteed ordered selection of futures. (However, this macro and stream require
+mapping all arguments to the same type.)
+
+The `futures::select` `Either` return type is complex, particularly when nested. This
+makes code hard to read and maintain. Map the `Either` to a custom enum.
+
+## Testing Async Code
+[testing-async-code]: #testing-async-code
+
+Zebra's existing acceptance and integration tests will catch most hangs and deadlocks.
+
+Some tests are only run after merging to `main`. If a recently merged PR fails on `main`,
+we revert the PR, and fix the failure.
+
+Some concurrency bugs only happen intermittently. Zebra developers should run regular
+full syncs to ensure that their code doesn't cause intermittent hangs. This is
+particularly important for code that modifies Zebra's highly concurrent crates:
+* `zebrad`
+* `zebra-network`
+* `zebra-state`
+* `zebra-consensus`
+* `tower-batch`
+* `tower-fallback`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Implementing and reviewing these constraints creates extra work for developers.
+But concurrency bugs slow down every developer, and impact users. And diagnosing
+those bugs can take a lot of developer effort.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+<!--
+- What makes this design a good design?
+- Is this design a good basis for later designs or implementations?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+-->
+
+# Prior art
+[prior-art]: #prior-art
+
+<!--
+Discuss prior art, both the good and the bad, in relation to this proposal.
+A few examples of what this can include are:
+
+- For community proposals: Is this done by some other community and what were their experiences with it?
+- For other teams: What lessons can we learn from what other communities have done here?
+- Papers: Are there any published papers or great posts that discuss this? If you have some relevant papers to refer to, this can serve as a more detailed theoretical background.
+
+This section is intended to encourage you as an author to think about the lessons from other projects, to provide readers of your RFC with a fuller picture.
+If there is no prior art, that is fine - your ideas are interesting to us whether they are brand new or if they are an adaptation from other projects.
+
+Note that while precedent set by other projects is some motivation, it does not on its own motivate an RFC.
+Please also take into consideration that Zebra sometimes intentionally diverges from common Zcash features and designs.
+-->
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Can we catch these bugs using automated tests?
+
+How can we diagnose these kinds of issues faster and more reliably?
```

Comment:
I'm very interested in these questions but don't have any leads to follow for them presently

Reply:
Tracing, spans, and metrics can really help, I'll mention them in the testing section.